rest_framework_json_api/renderers.py

```diff
@@ -382,6 +381,9 @@ def extract_root_meta(serializer, resource):
 
     @staticmethod
     def build_json_resource_obj(fields, resource, resource_instance, resource_name):
+        # Determine type from the instance if the underlying model is polymorphic
+        if isinstance(resource_instance, utils.POLYMORPHIC_ANCESTORS):
+            resource_name = utils.get_resource_type_from_instance(resource_instance)
```

Comment:
Afaik, all the `POLYMORPHIC_ANCESTORS` related code could be dropped if we can detect if the resource is polymorphic here.